hazelcast/src/test/java/com/hazelcast/internal/nearcache/AbstractBasicNearCacheTest.java

```diff
@@ -0,0 +1,420 @@
+import static com.hazelcast.config.EvictionPolicy.LRU;
+import static com.hazelcast.config.EvictionPolicy.NONE;
+import static com.hazelcast.config.InMemoryFormat.OBJECT;
+import static com.hazelcast.config.NearCacheConfig.LocalUpdatePolicy.CACHE;
+import static java.lang.String.format;
+import static java.util.concurrent.Executors.newFixedThreadPool;
+import static org.junit.Assert.assertEquals;
+import static org.junit.Assert.assertFalse;
+import static org.junit.Assert.assertNull;
+import static org.junit.Assert.assertTrue;
+
+/**
+ * Contains the logic code for Near Cache tests.
+ *
+ * @param <NK> key type of the tested Near Cache
+ * @param <NV> value type of the tested Near Cache
+ */
+public abstract class AbstractBasicNearCacheTest<NK, NV> extends HazelcastTestSupport {
+
+    /**
+     * The default count to be inserted into the Near Caches.
+     */
+    protected static final int DEFAULT_RECORD_COUNT = 1000;
+
+    /**
+     * The default name used for the data structures which have a Near Cache.
+     */
+    protected static final String DEFAULT_NEAR_CACHE_NAME = "defaultNearCache";
+
+    /**
+     * The {@link NearCacheConfig} used by the Near Cache tests.
+     *
+     * Needs to be set by the implementations of this class in their {@link org.junit.Before} methods.
+     */
+    protected NearCacheConfig nearCacheConfig;
+
+    /**
+     * Creates the {@link NearCacheTestContext} used by the Near Cache tests.
+     *
+     * @param <K> key type of the created {@link DataStructureAdapter}
+     * @param <V> value type of the created {@link DataStructureAdapter}
+     * @return a {@link NearCacheTestContext} used by the Near Cache tests
+     */
+    protected abstract <K, V> NearCacheTestContext<K, V, NK, NV> createContext();
+
+    protected final void populateMap(NearCacheTestContext<Integer, String, NK, NV> context) {
+        for (int i = 0; i < DEFAULT_RECORD_COUNT; i++) {
+            context.dataAdapter.put(i, "value-" + i);
+        }
+    }
+
+    protected final void populateNearCache(NearCacheTestContext<Integer, String, NK, NV> context) {
+        for (int i = 0; i < DEFAULT_RECORD_COUNT; i++) {
+            context.nearCacheAdapter.get(i);
+        }
+    }
+
+    @SuppressWarnings("unchecked")
+    private NK getNearCacheKey(NearCacheTestContext<Integer, String, NK, NV> context, int key) {
+        return (NK) context.serializationService.toData(key);
+    }
+
+    /**
+     * Checks that the Near Cache never returns its internal {@link NearCache#NULL_OBJECT} to the public API.
+     */
+    @Test
+    public void whenEmptyMap_thenPopulatedNearCacheShouldReturnNull_neverNULLOBJECT() {
+        NearCacheTestContext<Integer, String, NK, NV> context = createContext();
+
+        for (int i = 0; i < DEFAULT_RECORD_COUNT; i++) {
+            // populate Near Cache
+            assertNull(context.nearCacheAdapter.get(i));
+            // fetch value from Near Cache
```

Comment:
it's good to pass `i` to the error message, since when it fails there's no indication when it failed